External template api

NoumbissiValere opened the pull request.

nemesisdesign commented on openwisp_controller/config/models.py:
```diff
@@ -146,29 +151,6 @@ class Meta(AbstractTaggedTemplate.Meta):
-        self._validate_org_relation('vpn')
-        super(Template, self).clean()
-
-
```
why are you moving the `Template` model down? I would be very grateful if you avoid to do this kind of changes because it pollutes the PR and adds more work on the reviewer (me) to understand what is going on

nemesisdesign replied:
what about this comment? Can you follow this instruction please?

nemesisdesign commented on openwisp_controller/config/models.py:
```diff
@@ -198,6 +180,49 @@ def _auto_create_cert_extra(self, cert):
+        self._validate_org_relation('vpn')
+        if self.sharing == 'import':
+            data = self._get_remote_template_data()
+            import_org = Organization(**data['organization'])
```
I don't understand, why are you creating an organization??

NoumbissiValere replied:
@nemesisdesign this work was from another PR. it was needed here. it is not part of the logic to create external templates on the library. that is why it is executed for templates whose sharing setting is `import`. templates with `sharing settings = import` can't exist on the library. only public and shared_secret templates can. this was the import logic which admins will use on their different `openwisp_controller` instance to import the templates remotely.
I am creating the organization with which the template was imported with . that is if it doesn't already exist on that `openwisp_controller` instance.

nemesisdesign replied:
I still don't understand, I think we need to discuss this

nemesisdesign commented on openwisp_controller/config/models.py:
```diff
@@ -251,3 +276,23 @@ class Meta:
+    template = models.ForeignKey('config.Template',
+                                 verbose_name=_('Template'),
+                                 on_delete=models.CASCADE,
+                                 )
```
please indent the code properly

nemesisdesign commented on openwisp_controller/config/api/generics.py:
```diff
@@ -0,0 +1,82 @@
+    This API will be used at the template library
+    backend repo
+    """
+    authentication_classes = (SessionAuthentication, BasicAuthentication)
```
I think `BasicAuthentication` should be removed

nemesisdesign replied:
why is `BasicAuthentication` still here?

nemesisdesign commented on openwisp_controller/config/api/generics.py:
```diff
@@ -0,0 +1,82 @@
+        success = {
+            'template_success': 'Template was successfully created'
+        }
+        return Response(data=success, status=200)
```
this looks wrong:

- why is a VPN object being created?
- these operations should be done in a ModelSerializer

NoumbissiValere replied:
I had no idea this could be achieve with ModelSerializer. I will look how to achieve this with it. Thanks

nemesisdesign commented on openwisp_controller/config/api/urls.py:
```diff
@@ -0,0 +1,11 @@
+app_name = 'openwisp_controller'
+
+urlpatterns = [
+    url(r'^api/v1/templates/create/$',
```
the url path should be `r'^api/v1/templates/$'`, this URL must be able to do both listing for `GET` requests and create new templates for `POST` requests

NoumbissiValere commented:
Thanks for reviewing the work @nemesisdesign . will make the requested changes

NoumbissiValere commented:
@nemesisdesign i am very sorry about the poor indentation of code.

NoumbissiValere commented:
@nemesisdesign Please this is ready for review

NoumbissiValere commented:
@nemesisdesign . Please i will love that we discuss on this PR, this contains the notification model from PR #114 . I will have loved to make this notification only available  in `django_netjsonconfig` such that openwisp_controller doesn't even know it exist but at the background it is been used to notify admins. I did some research and some work but didn't succeed. I am saying so because, `TemplateSubscription` model has nothing to do with `organization` so i don't see a reason why it should be here. But if you are ok with the fact that this model is imported here, then it is ok. 

Also, I will love that you look at the Celery task. don't know if there is an error in importing it from `django_netjsonconfig`. I tried testing it here but it didn't work as expected.

Equally, You asked me to add a list feature to the API to create external templates on the template_library backend and change the URL from `api/v1/templates/create/$'` to `api/v1/templates`. 
I have added the feature to the API, but changing the url will cause a problem because there is already another base API in `django_netjsonconfig` which is incharge for listing and searching for public templates with this exact URL. i thought of moving this API to the `django_netjsonconfig` to eliminate this conflict or repetition, but i realized it won't be possible. reason being that the only way to restrict authenticated user to list and search for only their templates is to make use of the users field of the organization that the template belongs to. and `django_netjsonconfig` doesn't know about the concept of `organizations`. with this, i implemented the feature asked for but kept the url as `api/v1/templates/list-create/$` which will be used by authenticated users while the base API in `django_netjsonconfig` with url `api/v1/templates/$` will be used by unauthenticated users to search for templates which they wish to import. 

Following [this](https://www.django-rest-framework.org/api-guide/relations/#writable-nested-serializers), i will still need to create VPNs explicitly because `ModelSerializer` can't save nested serialization. I have done this is this PR.

nemesisdesign commented on openwisp_controller/config/api/views.py:
```diff
@@ -47,5 +49,25 @@ def get_queryset(self):
 template_detail = TemplateDetailView.as_view()
 list_template = ListTemplateView.as_view()
+create_template = ListCreateTemplateView.as_view()
+notify_template = TemplateSubscriptionView.as_view()
```
don't use the word `notify`

nemesisdesign commented on openwisp_controller/config/models.py:
```diff
@@ -251,3 +272,22 @@ class Meta:
+        abstract = False
+
+
+@periodic_task(run_every=60)
```
same concept discussed in `django-netjsonconfig` applies here

nemesisdesign commented on openwisp_controller/config/api/serializers.py:
```diff
@@ -41,3 +47,51 @@ class ListOrgTemplateSerializer(ListTemplateSerializer):
     pass
+
+
+class ListCreateTemplateSerializer(TemplateDetailSerializer, serializers.Field):
```
you may call this just `TemplateSerializer`, unless there are more template serializers

NoumbissiValere commented:
@nemesisdesign I need to send the Vpn object to the template to be used by the abstract class. reason why i need it to come before the template. else django will throw me an error saying am using `Vpn` before it is defined
